### DISCUSSION TITLE
App Router + Axios cause memory leak

### DISCUSSION THREAD

**Discussion by thinceller** [vercel/next.js#74307] [Help]
### Summary

Memory leak occurs when using Next.js App Router with Axios when Axios throws an error during React Server Component (RSC) rendering. The heap size keeps increasing with multiple requests and doesn't get released over time. This issue doesn't occur when using the fetch API instead of Axios.

For a detailed reproduction of this issue, please refer to our reproduction repository linked below.

### Environment
- Next.js 15.1.2
- React 19.0.0
- Node.js 20.18.1
- Axios 1.7.9

<details>

<summary>next info</summary>

```
Node.js v20.18.1

Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 24.1.0: Thu Oct 10 21:03:15 PDT 2024; root:xnu-11215.41.3~2/RELEASE_ARM64_T6000
  Available memory (MB): 65536
  Available CPU cores: 10
Binaries:
  Node: 20.18.1
  npm: 10.8.2
  Yarn: N/A
  pnpm: 9.15.1
Relevant Packages:
  next: 15.1.2 // Latest available version is detected (15.1.2).
  eslint-config-next: N/A
  react: 19.0.0
  react-dom: 19.0.0
  typescript: 5.5.4
Next.js Config:
  output: N/A
```

</details>

### Problem Details
- The heap size continuously increases as numerous requests are sent
- The increased memory is not released even after waiting for some time
- The memory leak seems to be associated with Axios throwing an error during React Server Component (RSC) rendering
- Switching from Axios to the fetch API resolves the memory leak issue and stabilizes the heap size
- This phenomenon has been observed in containerized environments, specifically on AWS ECS

### Expected Behavior
The heap size should stabilize after requests are completed, and unused memory should be released, even when errors occur.

### Actual Behavior
When Axios throws an error during RSC rendering, the heap size keeps increasing with each request and doesn't get released, leading to a memory leak.

### Example

https://github.com/thinceller/reproduction-app-router-axios-memory-leak

**Comment by justinadkins** [comment]
@thinceller We've observed a memory leak on 15.1.x in our application and we are using a dependency that uses Axios under the hood. I haven't had time to investigate but downgrading to 15.0.x has been our "hot fix". Do you happen to know if you are observing this issue on 15.0.x?

**Reply by thinceller** [reply]
@justinadkins I initially encountered this issue in production with Next.js 15.0.2. The problem still occurs when using 15.0.2 in the reproduction repo.

**Reply by justinadkins** [reply]
I see, well looks to be a separate issue then. We removed the dependency that was using Axios and trying to upgrade to 15.1.x and are still seeing a memory leak. Downgrading to 15.0.x however, resolves the issue.

<img width="574" alt="Screenshot 2025-01-06 at 9 06 48 AM" src="https://github.com/user-attachments/assets/721e8369-edfd-482b-8048-d977221a3ed4" />

**Comment by flameddd** [comment]
Environment
- Next.js 15.1.0 (Page Router)
- React 18.3.1
- Node.js 18.18.1
- Axios (Use in `/api`)


I also encountering memory leak issues. I cannot confirm whether it is related to Axios. Unfortunately, I am unable to investigate the problem further due to company policy.

Issue resolved by downgrading NextJS to `v14.2.22`

I did NOT test v15.0.x  


thanks everyone

**Reply by flameddd** [reply]
solved, ref: https://github.com/vercel/next.js/issues/74855

**Comment by wishket-pjw** [comment]
Environment

Next.js 15.3.1 (App Router)
React 19.1.0
Node.js 22.5.1
Axios 1.9.0 (Use in /api & RSC)

I have same issue.
This graph is production service memory trend.
![image](https://github.com/user-attachments/assets/332f2bde-3bd3-411e-a2b7-96dda7b4705c)
![image](https://github.com/user-attachments/assets/86e7de7b-7793-4250-a022-2e4cc26d4add)
(The memory is being initialized because I recreated the Task)

Downgrading to version 15.0.4 resolves the issue. 
However, this is not the direction I want.
Is there anyone who has solved this problem?

**Reply by wishket-pjw** [reply]
I think I solved the issue by upgrading my Node.js version.
I upgraded Node.js from version 22.5.1 to 23.6.1.
Now the graphs are rendering nicely.
![image](https://github.com/user-attachments/assets/ec247a0c-3db9-4637-867b-7326f5ca8b65) 
(800 users)
I think I need to monitor it a bit more though.

I referred to this reference:
https://github.com/vercel/next.js/issues/74855

**Reply by thinceller** [reply]
Thank you for sharing!

I'm using Next.js 15.3.1 and Node.js 22.14.0 (these versions should resolve the problem) in production env now, but it still has the memory leak.
I'll try other Node.js versions.

**Reply by louiswol94** [reply]
HI, did you solve this?
